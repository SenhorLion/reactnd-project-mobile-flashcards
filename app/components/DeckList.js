import React, { Component } from 'react';
import { View, Text, StyleSheet, ScrollView, FlatList } from 'react-native';
import { black } from '../utils/colors';
import AppModal from './ui/AppModal';
import Deck from './Deck';
import { DeleteModalConfirm } from './Modals';

class DeckList extends Component {
  state = {
    modalVisible: false,
    deckSelected: null,
  };

  handleDeleteDeck = deck => {
    console.log('DECK_LIST :: handleDeleteDeck', deck);

    this.setState({ deckSelected: deck }, () => {
      this.toggleModalVisible();
    });
  };

  confirmDeleteDeck = () => {
    const { deckSelected } = this.state;

    this.props.onDeleteDeck(deckSelected.id).then(res => {
      this.toggleModalVisible();
      this.clearSelectedDeck();
    });
  };

  clearSelectedDeck = () => {
    this.setState(prevState => ({
      deckSelected: null,
    }));
  };

  toggleModalVisible = () => {
    this.setState(prevState => ({
      modalVisible: !prevState.modalVisible,
    }));
  };

  closeModal = () => {
    this.setState({ modalVisible: false });
  };

  render() {
    const { decks } = this.props;
    const { isFetching, items } = decks;

    const hasDecks = !!(items && Object.keys(items).length);

    if (isFetching) {
      return (
        <View
          style={{
            flex: 6,
            alignItems: 'center',
            justifyContent: 'center',
          }}
        >
          <Text dataTestId="loading" style={{ fontSize: 20 }}>
            Loading Decks...
          </Text>;
        </View>
      );
    }

    if (!isFetching && !hasDecks) {
      return (
        <View
          style={{
            flex: 6,
            alignItems: 'center',
            justifyContent: 'center',
          }}
        >
          <Text dataTestId="loading" style={{ fontSize: 20 }}>
            No Decks available to show.
          </Text>;
        </View>
      );
    }

    return (
      <ScrollView style={{ flex: 1 }}>
        {hasDecks &&
          Object.values(items).map(deck => (
            <Deck
              key={deck.title}
              handleDeleteDeck={this.handleDeleteDeck}
              deck={deck}
              {...this.props}
            />
          ))}
        <AppModal
          backdropColor={black}
          isVisible={this.state.modalVisible}
          closeModal={this.closeModal}
          onBackdropPress={this.toggleModalVisible}
        >
          <DeleteModalConfirm
            title="Delete Deck"
            confirmDelete={this.confirmDeleteDeck}
            confirmCancel={this.toggleModalVisible}
          />
        </AppModal>
      </ScrollView>
    );
  }
}

const styles = StyleSheet.create({
  container: {
    flex: 6,
  },
});

export default DeckList;
