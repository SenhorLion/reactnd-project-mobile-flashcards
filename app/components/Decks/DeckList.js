import React, { Component } from 'react';
import { View, Text, ScrollView } from 'react-native';
import { black } from '../../utils/colors';
import AppModal from '../ui/AppModal';
import Deck from './Deck';
import { DeleteModalConfirm } from '../Modals';
import styles from './styles';

class DeckList extends Component {
  state = {
    modalVisible: false,
    deckSelected: null,
  };

  handleDeleteDeck = deck => {
    this.setState({ deckSelected: deck }, () => {
      this.toggleModalVisible();
    });
  };

  confirmDeleteDeck = () => {
    const { deckSelected } = this.state;

    this.props.onDeleteDeck(deckSelected.id).then(res => {
      this.toggleModalVisible();
      this.clearSelectedDeck();
    });
  };

  clearSelectedDeck = () => {
    this.setState(prevState => ({
      deckSelected: null,
    }));
  };

  toggleModalVisible = () => {
    this.setState(prevState => ({
      modalVisible: !prevState.modalVisible,
    }));
  };

  closeModal = () => {
    this.setState({ modalVisible: false });
  };

  render() {
    const { decks } = this.props;
    const { isFetching, items } = decks;

    const hasDecks = !!(items && Object.keys(items).length);

    if (isFetching) {
      return (
        <View
          style={{
            flex: 6,
            alignItems: 'center',
            justifyContent: 'center',
          }}
        >
          <Text dataTestId="loading" style={styles.messageText}>
            Loading Decks...
          </Text>
        </View>
      );
    }

    if (!isFetching && !hasDecks) {
      return (
        <View
          style={{
            flex: 6,
            alignItems: 'center',
            justifyContent: 'center',
          }}
        >
          <Text dataTestId="loading" style={styles.messageText}>
            No Decks available to show.
          </Text>
        </View>
      );
    }

    return (
      <ScrollView style={{ flex: 1 }}>
        {hasDecks &&
          Object.values(items).map(deck => (
            <Deck
              key={deck.title}
              handleDeleteDeck={this.handleDeleteDeck}
              deck={deck}
              {...this.props}
            />
          ))}
        <AppModal
          backdropColor={black}
          isVisible={this.state.modalVisible}
          closeModal={this.closeModal}
          onBackdropPress={this.toggleModalVisible}
        >
          <DeleteModalConfirm
            title="Delete Deck"
            confirmDelete={this.confirmDeleteDeck}
            confirmCancel={this.toggleModalVisible}
          />
        </AppModal>
      </ScrollView>
    );
  }
}

export default DeckList;
