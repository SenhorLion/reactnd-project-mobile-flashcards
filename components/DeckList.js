import React, { Component } from 'react';
import { View, Text, StyleSheet, ScrollView } from 'react-native';
import { AppLoading } from 'expo';
import { fetchDecksData } from '../api';
import {
  black,
  purple,
  lightPurple,
  white,
  gray,
  antiFlashWhite,
} from '../utils/colors';

import { enforcePromiseDelay } from '../utils/helpers';
import Deck from './Deck';
import ErrorBoundary from '../error/ErrorBoundary';

class DeckList extends Component {
  render() {
    const { decks, isDecksLoaded } = this.props;
    const hasDecks = !!(decks && Object.keys(decks).length);

    if (!isDecksLoaded) {
      return (
        <View
          style={{
            flex: 6,
            alignItems: 'center',
            justifyContent: 'center',
          }}
        >
          <Text dataTestId="loading" style={{ fontSize: 20 }}>
            Loading Decks...
          </Text>;
        </View>
      );
    }

    if (isDecksLoaded && !hasDecks) {
      return (
        <View
          style={{
            flex: 6,
            alignItems: 'center',
            justifyContent: 'center',
          }}
        >
          <Text dataTestId="loading" style={{ fontSize: 20 }}>
            No Decks available to show.
          </Text>;
        </View>
      );
    }

    return (
      <ScrollView style={{ flex: 1 }}>
        <ErrorBoundary>
          {hasDecks &&
            Object.keys(decks).map(deck => (
              <Deck
                key={decks[deck].title}
                deck={decks[deck]}
                {...this.props}
              />
            ))}
        </ErrorBoundary>
      </ScrollView>
    );
  }
}

// React: {
//     title: 'React',
//     questions: [
//       {
//         question: 'What is React?',
//         answer: 'A library for managing user interfaces',
//       },
//       {
//         question: 'Where do you make Ajax requests in React?',
//         answer: 'The componentDidMount lifecycle event',
//       },
//     ],
//   },

const styles = StyleSheet.create({
  container: {
    flex: 6,
  },
});

export default DeckList;
