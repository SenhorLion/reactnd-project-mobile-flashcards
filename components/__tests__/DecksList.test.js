import React from 'react';
import TestRenderer from 'react-test-renderer';
import DeckList from '../DeckList';

it('renders without crashing', () => {
  const DeckListComponent = TestRenderer.create(<DeckList />).toJSON();

  expect(DeckListComponent).toMatchSnapshot();
});

it('displays "loading" message while decks are not loaded', () => {
  const decksMock = {};
  const DeckListComponent = TestRenderer.create(
    <DeckList decks={decksMock} isDecksLoaded={false} />
  ).toJSON();

  expect(DeckListComponent).toMatchSnapshot();
});

it('displays "No Decks available" message if no decks exist', () => {
  const decksMock = {};
  const DeckListComponent = TestRenderer.create(
    <DeckList decks={decksMock} isDecksLoaded={true} />
  ).toJSON();

  expect(DeckListComponent).toMatchSnapshot();
});

it('displays all decks when decks are loaded and exist', () => {
  const decksMock = {
    React: {
      title: 'React',
      questions: [
        {
          question: 'What is React?',
          answer: 'A library for managing user interfaces',
        },
        {
          question: 'Where do you make Ajax requests in React?',
          answer: 'The componentDidMount lifecycle event',
        },
      ],
    },
  };
  const DeckListComponent = TestRenderer.create(
    <DeckList decks={decksMock} isDecksLoaded={true} />
  ).toJSON();
  expect(DeckListComponent).toMatchSnapshot();
});
