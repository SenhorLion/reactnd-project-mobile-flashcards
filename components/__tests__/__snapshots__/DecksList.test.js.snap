// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`displays "No Decks available" message if no decks exist 1`] = `
<View
  style={
    Object {
      "alignItems": "center",
      "flex": 6,
      "justifyContent": "center",
    }
  }
>
  <Text
    accessible={true}
    allowFontScaling={true}
    dataTestId="loading"
    ellipsizeMode="tail"
    style={
      Object {
        "fontSize": 20,
      }
    }
  >
    No Decks available to show.
  </Text>
  ;
</View>
`;

exports[`displays "loading" message while decks are not loaded 1`] = `
<View
  style={
    Object {
      "alignItems": "center",
      "flex": 6,
      "justifyContent": "center",
    }
  }
>
  <Text
    accessible={true}
    allowFontScaling={true}
    dataTestId="loading"
    ellipsizeMode="tail"
    style={
      Object {
        "fontSize": 20,
      }
    }
  >
    Loading Decks...
  </Text>
  ;
</View>
`;

exports[`displays all decks when decks are loaded and exist 1`] = `
<View
  style={
    Object {
      "alignItems": "center",
      "flex": 6,
      "justifyContent": "center",
    }
  }
>
  <Text
    accessible={true}
    allowFontScaling={true}
    dataTestId="loading"
    ellipsizeMode="tail"
    style={
      Object {
        "fontSize": 20,
      }
    }
  >
    No Decks available to show.
  </Text>
  ;
</View>
`;

exports[`renders without crashing 1`] = `
<View
  style={
    Object {
      "alignItems": "center",
      "flex": 6,
      "justifyContent": "center",
    }
  }
>
  <Text
    accessible={true}
    allowFontScaling={true}
    dataTestId="loading"
    ellipsizeMode="tail"
    style={
      Object {
        "fontSize": 20,
      }
    }
  >
    Loading Decks...
  </Text>
  ;
</View>
`;
